### PR TITLE
feat(data): add DataComp Energon preparation

### PR DESCRIPTION
**Pull request by cuichenx**
## Summary

- add a reproducible DataComp-1B download and audit tutorial based on the official downloader
- convert successful image-caption samples deterministically into indexed Energon WebDataset shards
- add a public `energon` dataset selector for compatible existing VLM recipes
- pin the Qwen processor revision and keep Energon micro-batch configuration synchronized with ordered recipe overrides

## Reference preparation

The documented 2026-07-24 preparation pinned the official DataComp repository,
the `mlfoundations/datacomp_1b` metadata revision, download-era dependencies,
and the Qwen3.6 processor revision.

- audited four metadata Parquets containing 2,133,334 URL rows
- produced 216 complete raw shards with 1,326,942 successful JPG/JSON/TXT triples
- fully decoded and deterministically converted 525,000 valid samples
- emitted 519,827 train and 5,173 validation samples in 53 indexed Energon tars
- loaded real train and validation batches through the pinned Qwen processor and assistant-only loss mask

The tutorial treats URL retrieval as fallible and gates conversion on audited
complete sample triples. The Qwen adaptation is causal image-conditioned
caption generation, not the canonical contrastive DataComp/CLIP objective.

## Training integration

The public workflow uses maintained conversion and training launchers. It
selects the existing Qwen3.5 35B recipe, which is architecture-compatible with
Qwen3.6, and explicitly sets MBS1/GBS512 for the image-caption run. No
Qwen3.6-specific recipe is added. Dataset-level micro-batch overrides remain
explicit and conflicting values continue to reach the existing validation
error.

The tutorial's 1,000-step command remains a worked example rather than a
claimed completed run. Downstream model card PR #5055 separately records
bounded DataComp projection-pretraining and checkpoint-resume evidence from a
clean integration commit containing this PR.

## Independent review

Independent reviews found and resolved two data-integrity edge cases. The
converter advertises only populated splits and rejects all-empty output before
indexing. The image downloader now revalidates every pinned metadata Parquet
immediately before consumption and rejects unexpected Parquet files so the
upstream wildcard cannot silently widen the selected slice. The final
post-fix review reported no remaining actionable findings.

## Resource safety

- CPU download, conversion, indexing, and tests request no GPUs
- the public tutorial does not request exclusive Slurm allocation
- the documented distributed import uses all eight requested GPUs at TP2/PP1/EP4
- training uses all eight requested GPUs at TP4/PP2/EP4

## Validation

- focused Energon builder, dataset-selector, and training-runner tests:
  161 passed
- tutorial shell syntax, metadata Python compilation, and unexpected-Parquet
  rejection smoke check: passed
- real Qwen processor batch reads: train and validation passed
- maintained conversion and training submission renders: passed
- full pre-commit hook set: passed
- `git diff --check` and public-path privacy scan: passed
- DCO sign-off: present

The direct tutorial unit-test file was removed at reviewer request. Conversion,
audit, metadata-manifest, and Energon-indexing behavior therefore retain
real-run evidence but no dedicated automated unit coverage in this PR.

### CODE REVIEW

File: scripts/training/run_recipe.py [hunk 1]
```diff
@@ -383,7 +383,8 @@ def _apply_dataset(recipe: ConfigContainer, args: argparse.Namespace) -> ConfigC
 
     recipe.dataset = build_dataset_config(recipe, args.dataset)
     requested_train_mode = _train_mode(args.mode)
-    if dataset_train_mode(recipe.dataset) != requested_train_mode:
+    required_train_mode = dataset_train_mode(recipe.dataset)
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py:386]: why need to change here?

Comment (cuichenx) [reply scripts/training/run_recipe.py:386]: i made "energon" format accepted for both 'pretrain' and 'finetune', so dataset_train_model returns None for energon. what do you think?

**Review by yaoyu-33 — APPROVED** [review]
